### ISSUE TITLE
Server side purge

### ISSUE THREAD

**Issue by garethbowen** [medic/medic#5443]
We recently added document purging on the client side which works really well for periodically deleting docs that are no longer needed. However because of the way it works it has to download the docs before it can determine if they should be purged. This will cause issues for people starting with a large number of docs, for example, if an existing user wipes the app or gets a new phone.

Instead, run the purge function on the server and store which users should purge the doc without deleting the doc itself - it must remain on the server. Existing clients will be told to purge the document from their local pouchdb. Exclude the document from the changes for those users so a new replication will never download it.

**Comment by garethbowen** [comment]
Ready for AT in `5443-server-side-purge`. The documentation is still being prepared but that can be done in parallel with the first rounds of AT. Work with @dianabarsan to determine how to configure and test this.

**Comment by dianabarsan** [comment]
Documentation PR: https://github.com/medic/medic-docs/pull/134

**Comment by newtewt** [comment]
Do we have an example purge that might be setup by a real life partner? I'm interested in running something similar on a copy of a real instance to see how it handles and how much space is taken up.

**Comment by dianabarsan** [comment]
We have 2 partners that have purge function setup (running locally). I used an optimized version of one of the functions for my benchmarks. I'll share details over Slack.

**Comment by newtewt** [comment]
@dianabarsan  does there need to be a medic-conf update for this? The current logic overwrites what is in the app_settings with the purging.js file. Meaning that if you were to define a  schedule to run in app_settings, then compile-app-settings in medic-conf it would overwrite and only have the purge function.

**Comment by dianabarsan** [comment]
@newtewt It looks like medic-conf is a bit uncertain as to where the purge settings should be:
https://github.com/medic/medic-conf/blob/master/src/fn/compile-app-settings.js#L70
It looks like it copies all settings from a `purging` property to the `purge` and then just updates the function. 
If you put your schedule config in `purging`, does it get the correct value in `purge` when compiled? 

It looks like medic-conf was initially bugged to create a `purging` property, while the app required `purge`. This quick patch (https://github.com/medic/medic-conf/commit/6f4cbd6ef83c202fa9919ab43f9587d9cbca80bb) changed it so the results would end up being stored under the correct property. 
Do you think we should fix this to have everything under the same property?

**Comment by newtewt** [comment]
Here's what happens. 

When I have my app_settings like this.  With a purging.js file
```
{
  "purging":{
    "text_expression": "every 1 mins"
  }
}
```
After running `medic-conf compile-app-settings ` it becomes
```
{
  "purging": {
    "text_expression": "every 1 mins",
    "fn": "function(userCtx, contact, reports, messages) {\n  const old = Date.now();\n  const oldMessages = Date.now();\n\n  const reportsToPurge = reports\n                           .filter(r => r.reported_date < old)\n                           .map(r => r._id);\n  const messagesToPurge = messages\n                           .filter(m => m.reported_date < oldMessages)\n                           .map(m => m._id);\n\n  return [...reportsToPurge, ...messagesToPurge];\n}\n"
  },
  "purge": {
    "text_expression": "every 1 mins",
    "fn": "function(userCtx, contact, reports, messages) {\n  const old = Date.now();\n  const oldMessages = Date.now();\n\n  const reportsToPurge = reports\n                           .filter(r => r.reported_date < old)\n                           .map(r => r._id);\n  const messagesToPurge = messages\n                           .filter(m => m.reported_date < oldMessages)\n                           .map(m => m._id);\n\n  return [...reportsToPurge, ...messagesToPurge];\n}\n"
  }
} 

```

I think we should probably get rid of the bugged property and only store under the correct property. I also know that Kenn has been trying to move things like tasks and targets to using the module.exports. I'm not sure if we could do that here but I think it would cut down on having to define a purge function in one place and add your scheduling in another.

**Comment by dianabarsan** [comment]
Ok, cool, I'll put it on tomorrow's TODO list :)

**Comment by dianabarsan** [comment]
@newtewt  Medic-conf update available on `5443-purge-instead-of-purging`. Please tell me if you're happy with it. Thanks!

**Comment by newtewt** [comment]
This is working as expected and merged!
